packages/server/src/http/resolveHTTPResponse.ts

```diff
@@ -41,72 +40,151 @@ interface ResolveHTTPRequestOptions<
   error?: Maybe<TRPCError>;
   contentTypeHandler?: BaseContentTypeHandler<any>;
   preprocessedBody?: boolean;
+  /**
+   * Called as soon as the response head is known,
+   * when streaming the response head will be determined
+   * **without** knowing the response body.
+   *
+   * Without this callback, streaming is disabled.
+   */
+  onHead?: (headResponse: Omit<HTTPResponse, 'body'>) => void;
+  /**
+   * Called for every procedure with `[index, result]`,
+   * or if not streaming (or error) a single time with `[-1, result]`
+   * where `result` is already stringified & transformed and ready to be sent.
+   *
+   * Without this callback, streaming is disabled.
+   */
+  onChunk?: (chunk: ResponseChunk) => void;
+}
+
+function initResponse<TRouter extends AnyRouter, TRequest extends HTTPRequest>(
+  ctx: inferRouterContext<TRouter> | undefined,
+  paths: string[] | undefined,
+  type:
+    | Exclude<(typeof HTTP_METHOD_PROCEDURE_TYPE_MAP)[string], undefined>
+    | 'unknown',
+  responseMeta?: HTTPBaseHandlerOptions<TRouter, TRequest>['responseMeta'],
+  untransformedJSON?:
+    | TRPCResponse<unknown, inferRouterError<TRouter>>
+    | TRPCResponse<unknown, inferRouterError<TRouter>>[]
+    | undefined,
+  errors: TRPCError[] = [],
```

Comment:
`max-params-2`

Also, this is a bit intense in the def - we can probably losen it a bit?

Reply:
Maybe it just handles too many cases. 

We should probably do ad-hoc ones for each code path rather than trying to reuse the code here, it's intense because it's so polymorphic

Reply:
This is not really complicated because it handles both streaming and regular. It's just complicated because I've taken it out of the long function body to isolate it in a separate function. The logic inside is the same as before, only the types are added, and sadly they are as minimal as can be to still work with `responseMeta`.